compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeNarrower.java

```diff
@@ -226,6 +232,17 @@ private NarrowedTypes getNarrowedTypesForBinaryOp(Map<BVarSymbol, NarrowedTypes>
             NarrowedTypes narrowedTypes = rhsTypes.get(symbol);
             rhsTrueType = narrowedTypes.trueType;
             rhsFalseType = narrowedTypes.falseType;
+            // Swapping the types if the RHS true type is a semantic error to disregard the RHS expr when continuing
+            // with the evaluation of the expression.
+            // e.g., ((x is int && x is boolean) && x is float) where x is of type int|string|float
+            // `x is boolean` will result in a semantic error since boolean is not in the above union. The false type
+            // of this type test contains the result from the last, semantically valid type test result. Since the
+            // true type is a semantic error, we swap the true and false types to use the last known correct narrowed
+            // type for future expr evaluations.
+            if (rhsTrueType.tag == TypeTags.SEMANTIC_ERROR && operator == OperatorKind.AND) {
```

Comment:
Is this only needed for `&&`?

What about something like

```ballerina
public function main() {
    int|string x = 1;

    if (x is int || x is int) && x == "abc" {

    }
}
```

Reply:
For this it gives the following error.. WDYT?
```
error: .::21609.bal:12:21: incompatible types: 'string' will not be matched to 'int'
```

Reply:
Yeah this is OK as discussed offline since we only get to the second case if the first is false.